SharkCage/CageNetwork/NetworkManager.cpp

```diff
@@ -5,76 +5,97 @@
-	case ExecutableType::SERVICE:
-		InitService();
-		break;
-	case ExecutableType::MANAGER:
-		InitManager();
-		break;
-	default:
-		break;
-	}
-}
+DLLEXPORT NetworkManager::NetworkManager(ContextType receive_for_type)
+{	
+	int receive_port = -1;
 
-DLLEXPORT std::wstring NetworkManager::Receive()
-{
-	rec_buf.clear();
-	rec_buf.resize(1024);
-	try
-	{
-		size_t len = socket.receive(asio::buffer(rec_buf));
-		rec_buf.resize(len);
-	}
-	catch (std::system_error e)
+	receive_port = GetPort(receive_for_type);
+
+	if (receive_port == -1)
 	{
-		std::wostringstream os;
-		os << e.what();
-		::OutputDebugString(os.str().c_str());
+		::OutputDebugString(L"except");
+		throw std::exception("Failed to initialize network manager");
 	}
-	return ToString(rec_buf);
-}
 
+	acceptor = std::make_unique<tcp::acceptor>(io_context, tcp::endpoint(tcp::v4(), receive_port));
+	acceptor->set_option(asio::ip::tcp::acceptor::reuse_address(true));
+}
 
-DLLEXPORT bool NetworkManager::Send(const std::wstring &msg)
+DLLEXPORT bool NetworkManager::Send(const std::wstring &msg, ContextType send_to_type)
 {
-	std::vector<char> message = ToCharVector(msg);
+	io_context.restart();
+
+	std::vector<char> message = StringToVec(msg);
 
 	try
 	{
-		tcp::socket tmp_socket(io_service);
-		tmp_socket.connect(send_endpoint);
+		tcp::resolver resolver(io_context);
+		tcp::resolver::query query(tcp::v4(), "localhost", std::to_string(GetPort(send_to_type)));
+		tcp::endpoint send_endpoint = *resolver.resolve(query);
+
+		tcp::socket socket(io_context);
+		socket.connect(send_endpoint);
 		send_buf = message;
-		tmp_socket.write_some(asio::buffer(send_buf));
-		tmp_socket.close();
+		socket.write_some(asio::buffer(send_buf));
 		return true;
 	}
 	catch (std::system_error e)
 	{
 		std::wostringstream os;
 		os << e.what();
 		::OutputDebugString(os.str().c_str());
-		return false;
 	}
+
+	return false;
 }
 
-DLLEXPORT std::wstring NetworkManager::Listen()
+DLLEXPORT std::wstring NetworkManager::Listen(long timeout_seconds)
 {
-	tcp::socket temp_socket(io_service);
-	acceptor.accept(temp_socket);
+	io_context.restart();
 
+	tcp::socket socket(io_context);
 	asio::streambuf buffer;
 
 	try
 	{
-		size_t len = asio::read_until(temp_socket, buffer, '\n');
+		asio::error_code err_code;
+		auto async_read_handler = [&err_code](const asio::error_code& ec, size_t)
+		{
+			err_code = ec;
+		};
+
+		auto async_accept_handler = [&socket, &buffer, &err_code, &async_read_handler](const asio::error_code& ec)
+		{
+			err_code = ec;
+			if (!ec)
+			{
+				asio::async_read_until(socket, buffer, '\n', async_read_handler);
+			}
+		};
+
+		acceptor->async_accept(socket, async_accept_handler);
+
+		if (timeout_seconds < 0)
+		{
+			io_context.run();
+		}
+		else
+		{
+			io_context.run_for(std::chrono::seconds::duration(timeout_seconds));
+
+			if (!io_context.stopped())
+			{
+				acceptor->cancel();
+				io_context.run();
+			}
+		}
+
+		if (err_code)
+		{
+			throw std::system_error(err_code);
```

Comment:
Am I complete confused or do you throw at this point an error, which you catch [three lines below](https://github.com/SharkCagey/HTWG_shark_cage/pull/52/files#diff-cc9ba72a5022ad84650f364c0ed35369R100)? If so, maybe it is better to write an function for printing the error and calling this function in the `if (err_code)` case as well as in the catch err?

Reply:
I don't think writing a function for 3 lines of code is an enhancement to be honest, this line just wraps an async error into an exception so all errors (the code above may throw other `system_erros`s) can be handled in the `catch`

Reply:
I thought about the (system) overhead creating an exception but year it should be ok so 👍

Reply:
https://xkcd.com/1691/ 😛